File: src/RepositoryModules.VB/RepositoryModule1.vb
Imports System.IO
Imports System.Text
Imports System.Threading
Imports Storex

<RepositoryModuleExport("E0B3F83A-B417-43DB-8CCF-9916A2EB63C6", "簡易ファイルシステム", Description:="モードでシンボル内容等を設定します")>
Public Class RepositoryModule1
    Implements IRepositoryModule

    Friend ReadOnly TextEncoding As Encoding = Encoding.GetEncoding("shift_jis")

    Private CurrentMode As IMode
    Private CurrentUser As IUser
    Private PrimaryLabelSpec As FixedLengthSpec
    Private SecondaryLabelCriteria As SecondaryLabelCriteria

    Private ConditionDictionary As IReadOnlyDictionary(Of String, IEnumerable(Of String))

    Public ReadOnly Property IsConfiguable As Boolean = True Implements IModule.IsConfiguable

    Public ReadOnly Property IsModeConfiguable As Boolean = True Implements IRepositoryModule.CanConfigureMode

    Public Function ConfigureAsync(cancellationToken As CancellationToken) As Task Implements IModule.ConfigureAsync

        ConfigForm1.Configure()

        Return Task.CompletedTask
    End Function

    Public Function ConfigureModeAsync(Mode As IMode, cancellationToken As CancellationToken) As Task Implements IRepositoryModule.ConfigureModeAsync

        ModeConfigForm1.Configure(Mode)

        Return Task.CompletedTask
    End Function

    Public Function PrepareAsync(Mode As IMode, UserInfo As IUser, cancellationToken As CancellationToken) As Task(Of Boolean) _
        Implements IRepositoryModule.PrepareAsync

        If Not File.Exists(MySettings.Default.FilePath) Then

            Throw New RepositoryModuleException("概要データ ファイルが正しく設定されていません。")
        End If

        If Not Directory.Exists(MySettings.Default.FolderPath) Then

            Throw New RepositoryModuleException("詳細データ フォルダが正しく設定されていません。")
        End If


        If Not Mode.TryExtractProperty(FixedLengthSpec.PropertyKeyForPrimary, PrimaryLabelSpec) _
            Or Not Mode.TryExtractProperty(SecondaryLabelCriteria.PropertyKey, SecondaryLabelCriteria) Then

            Throw New RepositoryModuleException("モードが構成されていません。")
        End If

        ConditionDictionary = ReadSecondaryConditionsFile()

        CurrentMode = Mode
        CurrentUser = UserInfo

        Return Task.FromResult(True)
    End Function

    Public Async Function FindPrimaryLabelAsync(Sources() As ILabelSource, cancellationToken As CancellationToken) As Task(Of ILabel) _
        Implements IRepositoryModule.FindPrimaryLabelAsync

        Await Task.CompletedTask

        Dim MatchSymbols = Sources _
            .OfType(Of Symbol) _
            .Select(
                Function(x)

                    Dim LabelDefinition As FixedLengthSpec = Nothing
                    Dim Label As BasicLabel = Nothing
                    Dim IsMatch = PrimaryLabelSpec.TryGenerateLabel(x, Label)

                    Return New With {IsMatch, Label}
                End Function
                ) _
            .Where(Function(x) x.IsMatch) _
            .Select(Function(x) x.Label) _
            .ToArray()

        Return If(MatchSymbols.Length = 1, MatchSymbols(0), Nothing)
    End Function

    Public Function FindSecondaryLabelsAsync(Primary As ILabel, Sources As ILabelSource(), cancellationToken As CancellationToken) As Task(Of ILabel()) _
        Implements IRepositoryModule.FindSecondaryLabelsAsync

        Dim Labels = New List(Of ILabel)()
        Dim Strings As IEnumerable(Of String) = Nothing

        Dim ValidConditionValues = If(SecondaryLabelCriteria.SpecifiedByConditionFileBehavior <> SecondaryLabelBehavior.Deny _
            AndAlso ConditionDictionary.TryGetValue(Primary.ItemNumber, Strings), Strings, Array.Empty(Of String)())

        Dim C3Labels = Sources _
            .Where(Function(__) (SecondaryLabelCriteria.AcceptableTypes And SecondaryLabelTypes.C3Label) = SecondaryLabelTypes.C3Label) _
            .OfType(Of C3Label)() _
            .Where(Function(o) _
                SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior <> SecondaryLabelBehavior.Deny _
                    AndAlso EqualsIgnoreCase(o.PartNumber, Primary.ItemNumber) _
                Or SecondaryLabelCriteria.SpecifiedByConditionFileBehavior <> SecondaryLabelBehavior.Deny _
                    AndAlso ValidConditionValues.Any(Function(p) EqualsIgnoreCase(o.PartNumber, p)) _
                Or SecondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior <> SecondaryLabelBehavior.Deny _
                    AndAlso Not EqualsIgnoreCase(o.PartNumber, Primary.ItemNumber) And Not ValidConditionValues.Any(Function(p) EqualsIgnoreCase(o.PartNumber, p))
                ) _
            .ToList()

        Labels.AddRange(C3Labels)

        Dim Label As ILabel = Nothing

        Dim SingleSymbolLabels = Sources _
            .Where(Function(__) (SecondaryLabelCriteria.AcceptableTypes And SecondaryLabelTypes.SingleSymbol) = SecondaryLabelTypes.SingleSymbol) _
            .OfType(Of Symbol)() _
            .Where(Function(x) Not PrimaryLabelSpec.TryGenerateLabel(x, Label)) _
            .Select(Function(o) New BasicLabel(o) With {.ItemNumber = o.Value}) _
            .Where(Function(o) _
                SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior <> SecondaryLabelBehavior.Deny _
                    And EqualsIgnoreCase(o.ItemNumber, Primary.ItemNumber) _
                Or SecondaryLabelCriteria.SpecifiedByConditionFileBehavior <> SecondaryLabelBehavior.Deny _
                    And ValidConditionValues.Any(Function(p) EqualsIgnoreCase(o.ItemNumber, p))
                ) _
            .ToList()

        Labels.AddRange(SingleSymbolLabels)

        Return Task.FromResult(Labels.ToArray())
    End Function

    Public Async Function RegisterAsync(Primary As ILabel, Secondary As ILabel, CaptureDatas As CaptureData(), Tags As String(), cancellationToken As CancellationToken) As Task(Of Boolean) _
        Implements IRepositoryModule.RegisterAsync

        If Not File.Exists(MySettings.Default.FilePath) Then
            Throw New RepositoryModuleException("概要データ ファイルが存在しません。")
        End If

        If Not Directory.Exists(MySettings.Default.FolderPath) Then
            Throw New RepositoryModuleException("詳細データ フォルダが存在しません。")
        End If

        Dim Strings As IEnumerable(Of String) = Nothing
        Dim ValidConditionValues = If(SecondaryLabelCriteria.SpecifiedByConditionFileBehavior <> SecondaryLabelBehavior.Deny _
            AndAlso ConditionDictionary.TryGetValue(Primary.ItemNumber, Strings), Strings, Array.Empty(Of String)())

        If Secondary Is Nothing Then

            If SecondaryLabelCriteria.NoLabelBehavior = SecondaryNoLabelBehavior.Warnning _
                AndAlso Not ShowAlert("セカンダリ ラベルが必要です。無視して登録しますか。") Then

                Return False

            ElseIf SecondaryLabelCriteria.NoLabelBehavior = SecondaryNoLabelBehavior.Deny Then

                ShowError("セカンダリ ラベルが必要です。")
                Return False

            ElseIf SecondaryLabelCriteria.NoLabelBehavior <> SecondaryNoLabelBehavior.Default Then

                Dim hasItemNumberEqualsToPrimary = Tags.Any(Function(o) EqualsIgnoreCase(o, Primary.ItemNumber))
                Dim hasSpecifiedByConditionFile = Tags.Any(Function(o) ValidConditionValues.Any(Function(p) EqualsIgnoreCase(p, Primary.ItemNumber)))

                If SecondaryLabelCriteria.NoLabelBehavior = SecondaryNoLabelBehavior.WarningWhenTagNotMatched _
                    AndAlso Not hasItemNumberEqualsToPrimary _
                    AndAlso Not hasSpecifiedByConditionFile _
                    AndAlso Not ShowAlert("セカンダリ ラベル、又は対応するタグが必要です。無視して登録しますか。") Then

                    Return False
                ElseIf SecondaryLabelCriteria.NoLabelBehavior = SecondaryNoLabelBehavior.DenyWhenTagNotMatched _
                    AndAlso Not hasItemNumberEqualsToPrimary _
                    AndAlso Not hasSpecifiedByConditionFile Then

                    ShowError("セカンダリ ラベル、又は対応するタグが必要です。")
                    Return False
                ElseIf SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior = SecondaryLabelBehavior.Warnning _
                    AndAlso hasItemNumberEqualsToPrimary _
                    AndAlso Not ShowAlert("プライマリラベルと同じ品番がタグに含まれています。登録しますか。") Then

                    Return False
                ElseIf SecondaryLabelCriteria.SpecifiedByConditionFileBehavior = SecondaryLabelBehavior.Warnning _
                    AndAlso hasSpecifiedByConditionFile _
                    AndAlso Not ShowAlert("対応表で指定されている品番がタグに含まれています。登録しますか。") Then

                    Return False
                End If
            End If
        ElseIf SecondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior = SecondaryLabelBehavior.Warnning _
            AndAlso EqualsIgnoreCase(Primary.ItemNumber, Secondary.ItemNumber) _
            AndAlso Not ShowAlert("指定された セカンダリ ラベルは、プライマリラベルと同じ品番です。登録を続行しますか。") Then

            Return False
        ElseIf SecondaryLabelCriteria.SpecifiedByConditionFileBehavior = SecondaryLabelBehavior.Warnning _
            AndAlso ValidConditionValues.Any(Function(o) EqualsIgnoreCase(o, Secondary.ItemNumber)) _
            AndAlso Not ShowAlert("指定された セカンダリ ラベルは、プライマリ ラベルとは異なり対応表と同じ品番です。登録を続行しますか。") Then

            Return False
        ElseIf SecondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior = SecondaryLabelBehavior.Warnning _
            AndAlso Not EqualsIgnoreCase(Primary.ItemNumber, Secondary.ItemNumber) _
            AndAlso Not ValidConditionValues.Any(Function(o) EqualsIgnoreCase(o, Secondary.ItemNumber)) _
            AndAlso Not ShowAlert("指定された セカンダリ ラベルは、プライマリラベルや対応表と異なる品番です。登録しますか。") Then

            Return False
        End If

        Dim Timestamp As Date = Date.Now

        Dim MyFile = New FileInfo(MySettings.Default.FilePath)

        Using SW = New StreamWriter(MyFile.FullName, True, TextEncoding)

            If MyFile.Length = 0 Then

                Await SW.WriteLineAsync(String.Join(vbTab, New String() {
                    "PrimaryPartNumber",
                    "PrimarySerialNumber",
                    "SecondaryPartNumber",
                    "SecondarySerialNumber",
                    "ModeId",
                    "UserName",
                    "Timestamp"
                    }))
            End If

            Await SW.WriteLineAsync(String.Join(vbTab, New String() {
                Primary.ItemNumber,
                Primary.SerialNumber,
                Secondary?.ItemNumber,
                Secondary?.SerialNumber,
                CurrentMode.Id.ToString(),
                CurrentUser?.DisplayName,
                $"{Timestamp:yyyy/MM/dd HH:mm:ss.fff}"
                }))
        End Using

        Dim MyFolder = New DirectoryInfo(MySettings.Default.FolderPath)

        MyFolder = MyFolder.CreateSubdirectory($"{Timestamp:yyyy-MM-dd HH-mm-ss.fff}")

        For i = 1 To CaptureDatas.Count

            Dim CaptureData As CaptureData = CaptureDatas(i - 1)

            Dim CaptureDataFolder = MyFolder.CreateSubdirectory($"catpure#{i}")

            WriteImage($"{CaptureDataFolder.FullName}\original.jpg", CaptureData.OriginalImage.ToArray())

            If CaptureData.AdornedImage IsNot Nothing Then

                WriteImage($"{CaptureDataFolder.FullName}\adorned.jpeg", CaptureData.AdornedImage.ToArray())
            End If

            If CaptureData.LabelSources.Count > 0 Then

                Dim SymbolValues() As String = CaptureData.LabelSources.SelectMany(Function(x) x.Symbols.Select(Function(xx) xx.Value)).ToArray()

                File.WriteAllLines($"{CaptureDataFolder.FullName}\symbols.txt", SymbolValues)
            End If
        Next

        If Tags.Length > 0 Then

            File.WriteAllLines($"{MyFolder.FullName}\tags.txt", Tags)
        End If

        Return True
    End Function

    Public Sub Dispose() Implements IDisposable.Dispose
    End Sub
End Class

File: src/RepositoryModules.CS/RepositoryModule1.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Storex;

namespace RepositoryModules
{
    [RepositoryModuleExport("E0B3F83A-B417-43DB-8CCF-9916A2EB63C6", "簡易ファイルシステム", Description = "モードでシンボル内容等を設定します")]
    public class RepositoryModule1 : IRepositoryModule
    {
        static readonly Encoding _textEncoding = Encoding.GetEncoding("shift_jis");

        IMode _currentMode;
        IUser _currentUser;

        public bool IsConfiguable { get; private set; } = true;

        public bool IsModeConfiguable { get; private set; } = true;

        public bool CanConfigureMode => true;

        public Task ConfigureAsync(CancellationToken cancellationToken)
        {
            ConfigForm1.Configure();

            return Task.CompletedTask;
        }

        public Task ConfigureModeAsync(IMode mode, CancellationToken cancellationToken)
        {
            ModeConfigForm1.Configure(mode);

            return Task.CompletedTask;
        }

        public Task<bool> PrepareAsync(IMode Mode, IUser UserInfo, CancellationToken cancellationToken)
        {
            if (!File.Exists(MySettings.Default.FilePath))
            {
                throw new RepositoryModuleException("概要データ ファイルが正しく設定されていません。");
            }

            if (!Directory.Exists(MySettings.Default.FolderPath))
            {
                throw new RepositoryModuleException("詳細データ フォルダが正しく設定されていません。");
            }

            if (!Mode.TryExtractProperty<FixedLengthSpec>(FixedLengthSpec.PropertyKeyForPrimary, out _priaryLabelFixedLengthSpec) |
                !Mode.TryExtractProperty<SecondaryLabelCriteria>(SecondaryLabelCriteria.PropertyKey, out _secondaryLabelCriteria))
            {
                throw new RepositoryModuleException("モードが構成されていません。");
            }

            _conditions = RepositoryModuleHelper.ReadSecondaryConditionsFile();

            _currentMode = Mode;
            _currentUser = UserInfo;

            return Task.FromResult(true);
        }

        FixedLengthSpec _priaryLabelFixedLengthSpec;

        public Task<ILabel> FindPrimaryLabelAsync(ILabelSource[] sources, CancellationToken cancellationToken)
        {
            var labels = sources
                .OfType<Symbol>()
                .Select(x => _priaryLabelFixedLengthSpec.TryGenerateLabel(x, out var label) ? label : null)
                .Where(x => x != null)
                .ToArray();

            return Task.FromResult(labels.Length == 1 ? labels[0] : null);
        }

        SecondaryLabelCriteria _secondaryLabelCriteria;
        IReadOnlyList<SecondaryCondition> _conditions;

        public Task<ILabel[]> FindSecondaryLabelsAsync(ILabel primary, ILabelSource[] sources, CancellationToken cancellationToken)
        {
            var labels = new List<ILabel>();

            var validConditionValues = _secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny ?
                _conditions
                    .Where(o => StringCompareHelper.CompareIgnoreCase(o.PrimaryLabelItemNumber, primary.ItemNumber))
                    .Select(o => o.SecondaryItemNumber)
                    .ToArray() :
                Array.Empty<string>();

            var c3Labels = sources
                .Where(_ => (_secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.C3Label) == SecondaryLabelTypes.C3Label)
                .OfType<C3Label>()
                .Where(o =>
                    (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior != SecondaryLabelBehavior.Deny && StringCompareHelper.CompareIgnoreCase(o.PartNumber, primary.ItemNumber))
                    | (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny && validConditionValues.Any(p => StringCompareHelper.CompareIgnoreCase(o.PartNumber, p)))
                    | (_secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior != SecondaryLabelBehavior.Deny && !StringCompareHelper.CompareIgnoreCase(o.PartNumber, primary.ItemNumber) & !validConditionValues.Any(p => StringCompareHelper.CompareIgnoreCase(o.PartNumber, p)))
                    )
                .ToList();

            labels.AddRange(c3Labels);

            var singleSymbolLabels = sources
                .Where(_ => (_secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.SingleSymbol) == SecondaryLabelTypes.SingleSymbol)
                .OfType<Symbol>()
                .Where(x => !_priaryLabelFixedLengthSpec.TryGenerateLabel(x, out _))
                .Select(o => new BasicLabel(o) { ItemNumber = o.Value })
                .Where(o =>
                    (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior != SecondaryLabelBehavior.Deny & StringCompareHelper.CompareIgnoreCase(o.ItemNumber, primary.ItemNumber))
                    | (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny & validConditionValues.Any(p => StringCompareHelper.CompareIgnoreCase(o.ItemNumber, p)))
                    )
                .ToList();

            labels.AddRange(singleSymbolLabels);

            return Task.FromResult(labels.ToArray());
        }

        public async Task<bool> RegisterAsync(ILabel primary, ILabel secondary, CaptureData[] captureDatas, string[] tags, CancellationToken cancellationToken)
        {
            if (!File.Exists(MySettings.Default.FilePath))
            {
                throw new RepositoryModuleException("概要データ ファイルが存在しません。");
            }

            if (!Directory.Exists(MySettings.Default.FolderPath))
            {
                throw new RepositoryModuleException("詳細データ フォルダが存在しません。");
            }

            if (secondary is null)
            {
                if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.Warnning
                    && !MessageHelper.ShowAlert("セカンダリ ラベルが必要です。無視して登録しますか。"))
                {
                    return false;
                }
                else if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.Deny)
                {
                    MessageHelper.ShowError("セカンダリ ラベルが必要です。");
                    return false;
                }
                else if (_secondaryLabelCriteria.NoLabelBehavior != SecondaryNoLabelBehavior.Default)
                {
                    var hasItemNumberEqualsToPrimary = tags.Any(o => StringCompareHelper.CompareIgnoreCase(o, primary.ItemNumber));
                    var hasSpecifiedByConditionFile = tags
                        .Any(o => _conditions
                            .Any(p => StringCompareHelper.CompareIgnoreCase(p.PrimaryLabelItemNumber, primary.ItemNumber)
                                & StringCompareHelper.CompareIgnoreCase(o, p.SecondaryItemNumber)
                                )
                            );

                    if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.WarningWhenTagNotMatched
                        && !hasItemNumberEqualsToPrimary && !hasSpecifiedByConditionFile
                        && !MessageHelper.ShowAlert("セカンダリ ラベル、又は対応するタグが必要です。無視して登録しますか。"))
                    {
                        return false;
                    }
                    else if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.DenyWhenTagNotMatched
                        && !hasItemNumberEqualsToPrimary && !hasSpecifiedByConditionFile)
                    {
                        MessageHelper.ShowError("セカンダリ ラベル、又は対応するタグが必要です。");
                        return false;
                    }
                    else if (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior == SecondaryLabelBehavior.Warnning
                        && hasItemNumberEqualsToPrimary
                        && !MessageHelper.ShowAlert("プライマリラベルと同じ品番がタグに含まれています。登録しますか。"))
                    {
                        return false;
                    }
                    else if (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior == SecondaryLabelBehavior.Warnning
                        && hasSpecifiedByConditionFile
                        && !MessageHelper.ShowAlert("対応表で指定されている品番がタグに含まれています。登録しますか。"))
                    {
                        return false;
                    }
                }
            }
            else if (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior == SecondaryLabelBehavior.Warnning
                && StringCompareHelper.CompareIgnoreCase(primary.ItemNumber, secondary.ItemNumber)
                && !MessageHelper.ShowAlert("指定された セカンダリ ラベルは、プライマリラベルと同じ品番です。登録を続行しますか。"))
            {
                return false;
            }
            else if (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior == SecondaryLabelBehavior.Warnning
                && _conditions.Any(o => StringCompareHelper.CompareIgnoreCase(o.PrimaryLabelItemNumber, primary.ItemNumber) & StringCompareHelper.CompareIgnoreCase(o.SecondaryItemNumber, secondary.ItemNumber))
                && !MessageHelper.ShowAlert("指定された セカンダリ ラベルは、プライマリ ラベルとは異なり対応表と同じ品番です。登録を続行しますか。"))
            {
                return false;
            }
            else if (_secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior == SecondaryLabelBehavior.Warnning
                && !StringCompareHelper.CompareIgnoreCase(primary.ItemNumber, secondary.ItemNumber)
                && !_conditions.Any(o => StringCompareHelper.CompareIgnoreCase(o.PrimaryLabelItemNumber, primary.ItemNumber) & StringCompareHelper.CompareIgnoreCase(o.SecondaryItemNumber, secondary.ItemNumber))
                && !MessageHelper.ShowAlert("指定された セカンダリ ラベルは、プライマリラベルや対応表と異なる品番です。登録しますか。"))
            {
                return false;
            }

            var timestamp = DateTime.Now;

            var myFile = new FileInfo(MySettings.Default.FilePath);

            using (var sw = new StreamWriter(myFile.FullName, true, _textEncoding))
            {
                if (myFile.Length == 0L)
                {
                    await sw.WriteLineAsync(string.Join("\t", new string[]
                    {
                        "PrimaryPartNumber",
                        "PrimarySerialNumber",
                        "SecondaryPartNumber",
                        "SecondarySerialNumber",
                        "ModeId",
                        "UserName",
                        "Timestamp",
                    }));
                }

                await sw.WriteLineAsync(string.Join("\t", new string[]
                {
                    primary.ItemNumber,
                    primary.SerialNumber,
                    secondary?.ItemNumber,
                    secondary?.SerialNumber,
                    _currentMode.Id.ToString(),
                    _currentUser?.DisplayName,
                    $"{timestamp:yyyy/MM/dd HH:mm:ss.fff}",
                }));
            }

            var mFolder = new DirectoryInfo(MySettings.Default.FolderPath);

            mFolder = mFolder.CreateSubdirectory($"{timestamp:yyyy-MM-dd HH-mm-ss.fff}");

            for (int i = 1, loopTo = captureDatas.Count(); i <= loopTo; i++)
            {
                var captureData = captureDatas[i - 1];

                var captureDataFolder = mFolder.CreateSubdirectory($"catpure#{i}");

                IOHelper.WriteImage($@"{captureDataFolder.FullName}\original.jpg", captureData.OriginalImage.ToArray());

                if (captureData.AdornedImage != null)
                {
                    IOHelper.WriteImage($@"{captureDataFolder.FullName}\adorned.jpeg", captureData.AdornedImage.ToArray());
                }

                if (captureData.LabelSources.Count > 0)
                {
                    var symbolValues = captureData.LabelSources.SelectMany(x => x.Symbols.Select(xx => xx.Value)).ToArray();

                    File.WriteAllLines($@"{captureDataFolder.FullName}\symbols.txt", symbolValues);
                }
            }

            if (tags.Length > 0)
            {
                File.WriteAllLines($@"{mFolder.FullName}\tags.txt", tags);
            }

            return true;
        }

        public void Dispose() { }
    }
}